spanner/cloudexecutor/executor/cloud_executor.go

```diff
@@ -0,0 +1,331 @@
+		t.tableColumnsByName[tableName] = make(map[string]*executorpb.ColumnMetadata)
+		for _, col := range table.GetColumn() {
+			t.tableColumnsByName[tableName][col.GetName()] = col
+		}
+	}
+}
+
+// getColumnType returns the column type of the given table and column.
+func (t *tableMetadataHelper) getColumnType(tableName string, colName string) (*spannerpb.Type, error) {
+	cols, ok := t.tableColumnsByName[tableName]
+	if !ok {
+		log.Printf("There is no metadata for table %s. Make sure that StartTransactionAction has TableMetadata correctly populated.", tableName)
+		return nil, spanner.ToSpannerError(status.Errorf(codes.InvalidArgument, "there is no metadata for table %s", tableName))
+	}
+	colMetadata, ok := cols[colName]
+	if !ok {
+		log.Printf("Metadata for table %s contains no column named %s", tableName, colName)
+		return nil, spanner.ToSpannerError(status.Errorf(codes.InvalidArgument, "metadata for table %s contains no column named %s", tableName, colName))
+	}
+	return colMetadata.GetType(), nil
+}
+
+// getColumnTypes returns a list of column types of the given table.
+func (t *tableMetadataHelper) getColumnTypes(tableName string) ([]*spannerpb.Type, error) {
+	cols, ok := t.tableColumnsInOrder[tableName]
+	if !ok {
+		log.Printf("There is no metadata for table %s. Make sure that StartTransactionAction has TableMetadata correctly populated.", tableName)
+		return nil, spanner.ToSpannerError(status.Errorf(codes.InvalidArgument, "there is no metadata for table %s", tableName))
+	}
+	var colTypes []*spannerpb.Type
+	for _, col := range cols {
+		colTypes = append(colTypes, col.GetType())
+	}
+	return colTypes, nil
+}
+
+// getKeyColumnTypes returns a list of key column types of the given table.
+func (t *tableMetadataHelper) getKeyColumnTypes(tableName string) ([]*spannerpb.Type, error) {
+	cols, ok := t.tableKeyColumnsInOrder[tableName]
+	if !ok {
+		log.Printf("There is no metadata for table %s. Make sure that StartTxnAction has TableMetadata correctly populated.", tableName)
+		return nil, fmt.Errorf("there is no metadata for table %s", tableName)
+	}
+	var colTypes []*spannerpb.Type
+	for _, col := range cols {
+		colTypes = append(colTypes, col.GetType())
+	}
+	return colTypes, nil
+}
+
+// outcomeSender is a utility class used for sending action outcomes back to the client. For read
+// actions, it buffers rows and sends partial read results in batches.
+type outcomeSender struct {
+	actionID int32
+	stream   executorpb.SpannerExecutorProxy_ExecuteActionAsyncServer
+
+	// partialOutcome accumulates rows and other relevant information
+	partialOutcome *executorpb.SpannerActionOutcome
+	readResult     *executorpb.ReadResult
+	queryResult    *executorpb.QueryResult
+
+	// All the relevant variables below should be set before first outcome is sent back,
+	// and unused variables should leave null.
+	timestamp              *timestamppb.Timestamp
+	hasReadResult          bool
+	hasQueryResult         bool
+	hasChangeStreamRecords bool
```

Comment:
Where is this one used?

Reply:
I will raise the changes for ChangeStream feature as a seperate PR.